client/client.go

```diff
@@ -1115,91 +1113,120 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+			return nil, fmt.Errorf("unable to generate key: %s", err)
 		}
-		return pubKeyList, nil
-	}
-
-	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
-		pubKeyList = make(data.KeyList, 0, 1)
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate key: %s", err)
-	}
-
-	// If a list of keys to rotate to are provided, we add those
-	if len(newKeys) > 0 {
-		pubKeyList = make(data.KeyList, 0, len(newKeys))
-		for _, keyID := range newKeys {
-			pubKey = r.CryptoService.GetKey(keyID)
-			if pubKey == nil {
+	case len(pubKeys) == 0 && len(privKeys) != 0:
+		for _, keyID := range privKeys {
+			if pubKey = r.CryptoService.GetKey(keyID); pubKey == nil {
 				return nil, fmt.Errorf("unable to find key: %s", keyID)
 			}
 			pubKeyList = append(pubKeyList, pubKey)
 		}
+	case len(pubKeys) != 0 && len(privKeys) != 0:
+		if err = privateKeyIDsMatchPublicKeys(privKeys, pubKeys); err != nil {
+			return nil, err
+		}
+		pubKeyList = pubKeys
+	default:
+		pubKeyList = pubKeys
 	}
 
-	// Convert to certs (for root keys)
-	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
-		return nil, err
+	return r.pubKeysToCerts(role, pubKeyList)
+}
+
+// privateKeyIDsMatchPublicKeys checks each KeyID to have a corresponding public key
+func privateKeyIDsMatchPublicKeys(privKIDs []string, pubKeys data.KeyList) error {
+	pubKeyIDs := make(map[string]bool)
+	for _, k := range pubKeys {
+		kid, err := utils.CanonicalKeyID(k)
+		if err != nil {
+			return fmt.Errorf("invalid public key: %v", err)
+		}
+		pubKeyIDs[kid] = true
 	}
 
-	return pubKeyList, nil
+	// check if each private key has corresponding public key
+	for _, id := range privKIDs {
+		if exist := pubKeyIDs[id]; !exist {
+			return fmt.Errorf("require a public key matching private key ID: %v", id)
+		}
+	}
+	return nil
 }
 
+// pubKeysToCerts converts public key to x509 public key
 func (r *NotaryRepository) pubKeysToCerts(role data.RoleName, pubKeyList data.KeyList) (data.KeyList, error) {
 	// only generate certs for root keys
 	if role != data.CanonicalRootRole {
 		return pubKeyList, nil
 	}
 
 	for i, pubKey := range pubKeyList {
-		privKey, loadedRole, err := r.CryptoService.GetPrivateKey(pubKey.ID())
+		kid, _ := utils.CanonicalKeyID(pubKey)
+		privKey, loadedRole, err := r.CryptoService.GetPrivateKey(kid)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error converting public key with id %v to private key: %v", kid, err)
 		}
 		if loadedRole != role {
 			return nil, fmt.Errorf("attempted to load root key but given %s key instead", loadedRole)
 		}
-		pubKey, err = rootCertKey(r.gun, privKey)
-		if err != nil {
-			return nil, err
+		// do nothing in case pubKey is already a x509 pubKey
+		if pubKey.Algorithm() != data.RSAx509Key && pubKey.Algorithm() != data.ECDSAx509Key {
```

Comment:
nit: can this check be a separate helper function ex:  `isx509Key`?  We have similar checking logic in `CanonicalKeyID` that could benefit from this too.  This can be a followup PR, though.